### PR TITLE
Get active element within the iframe when restoring focus

### LINKED ISSUE #67809: Block inserter does not work on widgets screen
### Description

I expect: Block list opens
Actual: Block list does not open

### Step-by-step reproduction instructions

1. Enable Theme TT1
2. Open Appearance > Widgets
3. Click the plus button in any widget area
4. Click [Browse all]

### Screenshots, screen recording, code snippet

https://github.com/user-attachments/assets/acbdf681-24b0-4bef-a715-ede98b6f21ea

### Environment info

WordPress: 6.7.1
Plugin: nothing
Theme: TT1

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [ ] Block
- [ ] Classic
- [x] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

### PR DESCRIPTION
**Pull request by jeryj**
Fixes https://github.com/WordPress/gutenberg/issues/67809

## What?
`useFocusReturn` doesn't work for items within the editor canvas iframe. Focus is sent to the `iframe` canvas instead of the active element within it. This PR allows for focus to get restored to the active element within the canvas.

## Why?

This root focus loss issue has caused a lot of fragile hotfixes. After this is merged, we can remove them. Here's a list of ones I know about:
- [Selecting the previous navigation link block when there is nothing to focus](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation-link/edit.js#L572-L614). This one may be trickier to address though, as the appender may disappear due to the whole "adding a temporary block and removing it" situation of the navigation block.
- [Always restoring focus when the block list appender quick inserter is closed or block selected](https://github.com/WordPress/gutenberg/pull/64877/files#diff-4cdcf916fb8cf5f88b98cd9e95ab532d90369cd768334f7896720a2b85c018bbR37)
- The above causes this issue: [Block inserter does not work on widgets screen](https://github.com/WordPress/gutenberg/issues/67809#top)
- In progress (hopefully will be fixed by removing the code that is itself a hotfix, rather than adding yet another hotfix on top of it): [Do not control Nav Link item focus when focus within sidebar control](https://github.com/WordPress/gutenberg/pull/68044#top)
- Probably others.

## How?

The `useFocusReturn` hook stores the `activeElement` as `focusedBeforeMount.current`, and then when popover/modal/etc closes, it returns focus to the `focusedBeforeMount.current`. 

This works well unless the `activeElement` is _within_ an `iframe` (like the editor canvas). If the element is within an `iframe` , then the `focusedBeforeMount.current` is set as the iframe itself, not the `activeElement` within the `iframe`.

This PR checks to see if the `activeElement` is an `iframe` and if the `iframe` has an `activeElement` itself. If so, it sets the `activeElement` within the `iframe` as the `focusedBeforeMount.current`.

## Testing Instructions

- On the iframed block editor
- Add a group block
- Click the appender
- Press Escape
- Focus should be on the appender again
- Press Enter to open the inserter again
- Repeat test with the social icon block appender

## Screenshots or screencast 

|Before|After|
|-|-|
|||

### CODE REVIEW

File: packages/block-editor/src/components/button-block-appender/index.js [hunk 1]
```diff
@@ -34,7 +30,6 @@ function ButtonBlockAppender(
 				if ( onSelect && typeof onSelect === 'function' ) {
 					onSelect( ...args );
 				}
-				inserterButtonRef.current?.focus();
```
Comment (jeryj) [inline packages/block-editor/src/components/button-block-appender/index.js:37]: This file is undoing work from https://github.com/WordPress/gutenberg/pull/64877/files#diff-4cdcf916fb8cf5f88b98cd9e95ab532d90369cd768334f7896720a2b85c018bbR37 which fixed moving focus to the appender, but did it without addressing the root issue.

File: packages/compose/src/hooks/use-focus-return/index.js [hunk 2]
```diff
@@ -48,7 +48,21 @@ function useFocusReturn( onFocusReturn ) {
...
+			// not the active element within the iframe.
+			if (
+				activeElement?.tagName === 'IFRAME' &&
+				activeElement.contentDocument?.activeElement
+			) {
+				focusedBeforeMount.current =
+					activeElement.contentDocument.activeElement;
+			} else {
+				focusedBeforeMount.current = activeElement;
+			}
```
Comment (ajlende) [inline packages/compose/src/hooks/use-focus-return/index.js]: I'd rephrase as something like this.

```suggestion
			const activeDocument =
				node.ownerDocument.activeElement instanceof
				window.HTMLIFrameElement
					? node.ownerDocument.activeElement.contentDocument
					: node.ownerDocument;

			focusedBeforeMount.current = activeDocument?.activeElement ?? null;
```

**Review by ajlende — APPROVED** [review]
I had the thought that this could be recursive for nested iframes, but we probably don't need that for now.

Just had one suggestion for cleaner, more robust code.

**Comment by t-hamano** [comment]
Thanks for the fix!

Would it be worth backporting this PR to WP 6.7.2? Because #67809 is an issue that first appeared in WP 6.7.

**Comment by jeryj** [comment]
> Would it be worth backporting this PR to WP 6.7.2? Because https://github.com/WordPress/gutenberg/issues/67809 is an issue that first appeared in WP 6.7.

I think so, provided no bugs pop up as a result of it :)

**Comment by aaronjorbin** [comment]
@jeryj Have you seen any bugs since this shipped in Gutenebrg 20.0 that would make you 2nd guess backporting for WP 6.7.2?

**Comment by jeryj** [comment]
I have not seen any, however, I was also on vacation and then working on other projects. So, I have not really used the editor much to test.

**Comment by t-hamano** [comment]
This fix was backported to WP 6.7.2 as #68200.
